File: spec/services/forms/visible_path_spec.rb
require 'rails_helper'

RSpec.describe Forms::VisiblePath do
  # Anamnese real: se está grávida, encaminha; se nunca fez procedimento,
  # salta a pergunta sobre quais foram.
  def schema(logics: [])
    {
      'sections' => [
        {
          'key' => 'saude',
          'fields' => [
            { 'key' => 'nome', 'type' => 'text', 'label' => 'Nome' },
            { 'key' => 'gravida', 'type' => 'select', 'label' => 'Grávida?', 'options' => %w[Sim Não] },
            { 'key' => 'ja_fez', 'type' => 'select', 'label' => 'Já fez procedimento?', 'options' => %w[Sim Não] },
            { 'key' => 'quais', 'type' => 'textarea', 'label' => 'Quais?' },
            { 'key' => 'alergias', 'type' => 'textarea', 'label' => 'Alergias' }
          ]
        }
      ],
      'endings' => [{ 'key' => 'encaminhar', 'label' => 'Vamos falar consigo' }],
      'logics' => logics
    }
  end

  def regra(owner:, ref:, comparison:, expected:, target:)
    { 'field_key' => owner,
      'payloads' => [{ 'condition' => { 'ref' => ref, 'comparison' => comparison, 'expected' => expected },
                       'action' => { 'kind' => 'navigate', 'field_key' => target } }] }
  end

  it 'walks every question when there is no logic' do
    percurso = described_class.new(schema: schema, answers: {})

    expect(percurso.keys).to eq(%w[nome gravida ja_fez quais alergias])
  end

  it 'skips what the jump leaves behind' do
    logics = [regra(owner: 'ja_fez', ref: 'ja_fez', comparison: 'is', expected: 'Não', target: 'alergias')]
    percurso = described_class.new(schema: schema(logics: logics), answers: { 'ja_fez' => 'Não' })

    # `quais` não foi escondida por uma condição sua: ficou pelo caminho.
    expect(percurso.keys).to eq(%w[nome gravida ja_fez alergias])
    expect(percurso).not_to be_visible('quais')
  end

  it 'walks straight through when the condition does not hold' do
    logics = [regra(owner: 'ja_fez', ref: 'ja_fez', comparison: 'is', expected: 'Não', target: 'alergias')]
    percurso = described_class.new(schema: schema(logics: logics), answers: { 'ja_fez' => 'Sim' })

    expect(percurso.keys).to eq(%w[nome gravida ja_fez quais alergias])
  end

  it 'ends the form when the jump points at an ending' do
    logics = [regra(owner: 'gravida', ref: 'gravida', comparison: 'is', expected: 'Sim', target: 'encaminhar')]
    percurso = described_class.new(schema: schema(logics: logics), answers: { 'gravida' => 'Sim' })

    expect(percurso.keys).to eq(%w[nome gravida])
  end

  it 'uses the first rule that holds, and ignores the ones after it' do
    logics = [{ 'field_key' => 'ja_fez',
                'payloads' => [
                  { 'condition' => { 'ref' => 'ja_fez', 'comparison' => 'is_not_empty', 'expected' => nil },
                    'action' => { 'kind' => 'navigate', 'field_key' => 'alergias' } },
                  { 'condition' => { 'ref' => 'ja_fez', 'comparison' => 'is', 'expected' => 'Sim' },
                    'action' => { 'kind' => 'navigate', 'field_key' => 'quais' } }
                ] }]
    percurso = described_class.new(schema: schema(logics: logics), answers: { 'ja_fez' => 'Sim' })

    expect(percurso.keys).to eq(%w[nome gravida ja_fez alergias])
  end

  it 'refuses to jump backwards instead of looping forever' do
    logics = [regra(owner: 'quais', ref: 'quais', comparison: 'is_not_empty', expected: nil, target: 'nome')]
    percurso = described_class.new(schema: schema(logics: logics), answers: { 'quais' => 'toxina' })

    expect(percurso.keys).to eq(%w[nome gravida ja_fez quais alergias])
  end

  it 'ignores a calculation when deciding the path' do
    logics = [{ 'field_key' => 'ja_fez',
                'payloads' => [{ 'condition' => { 'ref' => 'ja_fez', 'comparison' => 'is', 'expected' => 'Sim' },
                                 'action' => { 'kind' => 'calculate', 'variable' => 'risco',
                                               'operator' => 'addition', 'value' => '2' } }] }]
    percurso = described_class.new(schema: schema(logics: logics), answers: { 'ja_fez' => 'Sim' })

    # Calcular não muda o caminho; só somar a uma variável oculta.
    expect(percurso.keys).to eq(%w[nome gravida ja_fez quais alergias])
  end

  describe 'versões já publicadas' do
    it 'still honours the old visible_when' do
      antigo = {
        'sections' => [{ 'key' => 's', 'fields' => [
          { 'key' => 'ja_fez', 'type' => 'select', 'label' => 'Já fez?', 'options' => %w[Sim Não] },
          { 'key' => 'quais', 'type' => 'textarea', 'label' => 'Quais?',
            'visible_when' => { 'field' => 'ja_fez', 'operator' => 'equals', 'value' => 'Sim' } }
        ] }]
      }

      # As versões são imutáveis e continuam a receber respostas: partir isto
      # esconderia perguntas de formulários que já estão no ar.
      expect(described_class.new(schema: antigo, answers: { 'ja_fez' => 'Sim' }).keys).to eq(%w[ja_fez quais])
      expect(described_class.new(schema: antigo, answers: { 'ja_fez' => 'Não' }).keys).to eq(%w[ja_fez])
    end
  end
end

File: app/javascript/shared/helpers/specs/forms/visiblePath.spec.js
import { visiblePathKeys } from 'shared/helpers/forms/visiblePath';
import { matches, satisfied } from 'shared/helpers/forms/logic';

const schema = (fields, logics = []) => ({
  sections: [{ key: 's', fields }],
  logics,
});

const field = (key, extra = {}) => ({
  key,
  type: 'text',
  label: key,
  ...extra,
});

describe('visiblePathKeys', () => {
  it('walks every question when no rule interferes', () => {
    const path = visiblePathKeys(
      schema([field('a'), field('b'), field('c')]),
      {}
    );

    expect(path).toEqual(['a', 'b', 'c']);
  });

  it('skips what a jump leaves behind, not only what a condition names', () => {
    // É a diferença entre esconder e saltar: `b` não é mencionada por nenhuma
    // condição e mesmo assim não foi vista.
    const path = visiblePathKeys(
      schema(
        [field('a'), field('b'), field('c')],
        [
          {
            field_key: 'a',
            payloads: [
              {
                condition: { ref: 'a', comparison: 'is', expected: 'saltar' },
                action: { kind: 'navigate', field_key: 'c' },
              },
            ],
          },
        ]
      ),
      { a: 'saltar' }
    );

    expect(path).toEqual(['a', 'c']);
  });

  it('follows the first rule that holds and ignores the rest', () => {
    const logics = [
      {
        field_key: 'a',
        payloads: [
          {
            condition: { ref: 'a', comparison: 'is', expected: 'x' },
            action: { kind: 'navigate', field_key: 'b' },
          },
          {
            condition: { ref: 'a', comparison: 'is_not_empty' },
            action: { kind: 'navigate', field_key: 'c' },
          },
        ],
      },
    ];
    const path = visiblePathKeys(
      schema([field('a'), field('b'), field('c')], logics),
      { a: 'x' }
    );

    expect(path).toEqual(['a', 'b', 'c']);
  });

  it('ends the walk when a rule jumps to an ending', () => {
    const path = visiblePathKeys(
      schema(
        [field('a'), field('b')],
        [
          {
            field_key: 'a',
            payloads: [
              {
                condition: { ref: 'a', comparison: 'is_not_empty' },
                action: { kind: 'navigate', field_key: 'obrigado' },
              },
            ],
          },
        ]
      ),
      { a: 'sim' }
    );

    expect(path).toEqual(['a']);
  });

  it('never walks backwards, so two rules pointing at each other still end', () => {
    const logics = [
      {
        field_key: 'b',
        payloads: [
          {
            condition: { ref: 'b', comparison: 'is_not_empty' },
            action: { kind: 'navigate', field_key: 'a' },
          },
        ],
      },
    ];
    const path = visiblePathKeys(schema([field('a'), field('b')], logics), {
      b: 'volta',
    });

    expect(path).toEqual(['a', 'b']);
  });

  it('still honours a condition from an already published version', () => {
    // Versões publicadas são imutáveis e trazem `visible_when` com `equals`.
    const fields = [
      field('a'),
      field('b', {
        visible_when: { field: 'a', operator: 'equals', value: 'sim' },
      }),
    ];

    expect(visiblePathKeys(schema(fields), { a: 'nao' })).toEqual(['a']);
    expect(visiblePathKeys(schema(fields), { a: 'sim' })).toEqual(['a', 'b']);
  });
});

describe('logic comparators', () => {
  it('compares numbers as numbers, and refuses what is not one', () => {
    expect(matches('greater_than', '12', '5')).toBe(true);
    expect(matches('greater_than', '12abc', '5')).toBe(false);
    expect(matches('less_or_equal_than', '3,5', '4')).toBe(true);
  });

  it('treats a multiple selection as a set for `is` and a member for `contains`', () => {
    expect(matches('is', ['b', 'a'], ['a', 'b'])).toBe(true);
    expect(matches('is', ['a'], ['a', 'b'])).toBe(false);
    expect(matches('contains', ['a', 'b'], 'b')).toBe(true);
  });

  it('reads an unchecked consent as empty', () => {
    expect(matches('is_empty', false)).toBe(true);
    expect(matches('is_not_empty', true)).toBe(true);
  });

  it('combines conditions with and/or, one level deep', () => {
    const group = {
      combinator: 'any',
      conditions: [
        { ref: 'x', comparison: 'is', expected: '1' },
        { ref: 'y', comparison: 'is', expected: '2' },
      ],
    };

    expect(satisfied(group, { y: '2' })).toBe(true);
    expect(satisfied({ ...group, combinator: 'all' }, { y: '2' })).toBe(false);
  });

  it('never fires an empty group', () => {
    expect(satisfied({ combinator: 'all', conditions: [] }, {})).toBe(false);
  });

  it('returns false for an operator it does not know, instead of throwing', () => {
    expect(matches('inventado', 'a', 'a')).toBe(false);
  });
});
